cloud/whitelist.go

```diff
@@ -0,0 +1,60 @@
+// Copyright 2019 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+package cloud
+
+import (
+	"fmt"
+	"sort"
+	"strings"
+
+	"github.com/juju/collections/set"
+	"github.com/juju/errors"
+
+	"github.com/juju/juju/provider/lxd/lxdnames"
+)
+
+// WhiteList contains a cloud compatibility matrix:
+// if controller was bootstrapped on a particular cloud type,
+// what other cloud types can be added to it.
+type WhiteList struct {
+	matrix map[string]set.Strings
+}
+
+// String constructs user friendly white list representation.
+func (w *WhiteList) String() string {
+	if len(w.matrix) == 0 {
+		return "empty whitelist"
+	}
+	sorted := []string{}
+	for one := range w.matrix {
+		sorted = append(sorted, one)
+	}
+	sort.Strings(sorted)
+	result := []string{}
+	for _, one := range sorted {
+		result = append(result, fmt.Sprintf(" - controller cloud type %q supports %v", one, w.matrix[one].SortedValues()))
+	}
+	return strings.Join(result, "\n")
+}
+
+// Check will err out if 'exiting' controller cloud type is
+// not compatible with a 'new' cloud type according to this white list.
+func (w *WhiteList) Check(existing, new string) error {
+	if list, ok := w.matrix[existing]; ok {
+		if !list.Contains(new) {
+			return errors.Errorf("cloud type %q is not whitelisted for controller cloud type %q, current whitelist: %v", new, existing, list.SortedValues())
```

Comment:
"whitelist" might be too technical for some users.
Fine to return this message here though, and the final end user message should be rendered by the caller (see other comment), something like

"adding a cloud of type %q might not function correctly on a controller bootstrapped on a %q cloud; if you really want to do this use --force"

Reply:
I will have to change the wording here since apiserver call returns error not typed error.

This is fine since the actual whitelist is still visible in command help.

Reply:
Actually, params.Error can be returned since it complies with error interface. I'll change the code as per your suggestion. However, even though I like the message suggestion, it'll have to change slightly since we do not have bootstrapped controller's cloud at cmd here.